BootloaderCommonPkg/Library/ContainerLib/ContainerLib.c

```diff
@@ -246,6 +271,98 @@ AuthenticateComponent (
   return Status;
 }
 
+/**
+  This function registers a container.
+
+  @param[in]  ContainerBase      Container base address to register.
+
+  @retval EFI_NOT_READY          Not ready for register yet.
+  @retval EFI_BUFFER_TOO_SMALL   Insufficant max container entry number.
+  @retval EFI_OUT_OF_RESOURCES   No space to add new container.
+  @retval EFI_SUCCESS            The container has been registered successfully.
+
+**/
+EFI_STATUS
+RegisterContainer (
+  IN  UINT32   ContainerBase
+  )
+{
+  EFI_STATUS                Status;
+  CONTAINER_HDR            *ContainerHdr;
+  CONTAINER_ENTRY          *ContainerEntry;
+  COMPONENT_ENTRY          *CompEntry;
+  UINT8                    *AuthData;
+  UINT32                    ContainerHdrSize;
+  UINT8                     AuthType;
+  UINT64                    SignatureBuffer;
+  UINT8                    *DataBuf;
+  UINT8                    *CompData;
+  UINT32                    DataLen;
+  UINT32                    SignedDataLen;
+  UINT32                    Index;
+  LOADER_COMPRESSED_HEADER *CompressHdr;
+
+  ContainerHdr     = (CONTAINER_HDR *)ContainerBase;
+  SignatureBuffer  = ContainerHdr->Signature;
+  DEBUG ((EFI_D_INFO, "Registering container %4a\n", (CHAR8 *)&SignatureBuffer));
+
+  // Register container temporarily
+  Status = RegisterContainerInternal (ContainerBase);
+  if (!EFI_ERROR (Status)) {
```

Comment:
RegisterContainerInternal() is a private function, and others need not call this. Only RegisterContainer() is exposed outside. Based on the internal register return value, we decide to authenticate the component or not. But internal needs to deal with only registering, not authenticating.

WIll change the comment , temporarily means, if theres an error, we might unregister it soon.

Reply:
I think reg() and de-reg() is not good approach in failure case. What about authenticating first, then reg it if there is no error? In that case, de-reg() is not required.

Reply:
Unregister() is still required. Imagine there are 2 PODS with same container signature. We can unregister first one after we load the compinents. In that way, registration and loading other container doesnt throw error.

Reply:
I would agree reg/de-reg might not be the best way to handle it in this case.  But by doing so, it makes the code simpler to handle since all functions rely on the GetContainerPtr() pointer to get container info.  Before registration, the info cannot be retrieved.   We can for sure do the other way, but need more code to deal with it.  Since only RegisterContianer() is external API.  I think as long as we handle it carefully inside this lib,  it should be fine.  We can put some notes there if required.

Reply:
Understood. For code readability, what about having separate AuthenticateInternal()?
RegisterContainer() {
  RegisterContainerInternal()
  AuthenticateInternal()
  UnregisterContainer()
}
One more thing, if PcdVerifiedBootEnabled is disabled, AuthenticateComponent() returns EFI_NOT_FOUND. Need to handle disabled VerifiedBoot properly.

Reply:
for POD image (just like IAS image), do we need call register? I have thought only container that have sub-region need register.

Reply:
@myguitar   Good point.  We can rename the function to make it easier to understand.

Reply:
@gdong1   We treat POD in the same way as container.  There is no difference to the library.   I think it is better to do things in a consistent way.